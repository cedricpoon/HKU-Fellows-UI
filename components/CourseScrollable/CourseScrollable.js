import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as Animatable from 'react-native-animatable';
import { Text, Accordion, View, Icon as NBIcon } from 'native-base';
import { FlatList } from 'react-native';

import CourseLink from './CourseLink/CourseLink';
import styles from './Styles';
import {
  getCoursePathByIndex,
  getCoursePathById,
  getIndexByBreadcrumb
} from './helper';

const Icon = Animatable.createAnimatableComponent(NBIcon);

const bouncyIconDuration = 1500;

class CourseScrollable extends Component {

  constructor(props) {
    super(props);
    this.state = { mounted: false }
  }

  collapseAll = () => {
    this._rootAccordion.setSelected(-1);
  }

  _setSelectedWrapper = (ref, pid) => {
    const { list, onSetSelectCourseIndex } = this.props;

    return (index) => {
      /* cloned from Accordion */
      if (ref.state.selected === index) {
        ref.setState({ selected: undefined });
        onSetSelectCourseIndex(getCoursePathById(list, pid));
      } else {
        ref.setState({ selected: index });
        if (index !== -1)
          onSetSelectCourseIndex(getCoursePathByIndex(list, pid, index));
        else
          onSetSelectCourseIndex([]);
      }
    };
  }

  _renderHeader = ({title, inner}, expanded) => {
    const activeStyle = inner ? styles.active : styles.exoActive;

    return (
      <View
        style={styles.header}
      >
        <Text
          style={[
            styles.title,
            styles.headerText,
            expanded && activeStyle
          ]}
          numberOfLines={1}
        >
          {title}
        </Text>
        {expanded
          ? <Icon
              animation="rubberBand"
              duration={bouncyIconDuration}
              iterationCount="infinite"
              style={[styles.headerText, activeStyle]}
              name="arrow-up"
            />
          : <Icon style={styles.headerText} name="arrow-down" />}
      </View>
    );
  }

  _renderContent = ({children, links, id}) => {
    const { onItemPressWrapper, list, expandedList } = this.props;
    const { mounted } = this.state;

    if (children) {
      children.forEach((elem) => {
        elem.inner = true
      });
    }

    return (
      <View>
        {links && (
          <FlatList
            data={links}
            /* first link id as list key */
            listKey={links[0].id}
            keyExtractor={item => item.id}
            renderItem={({item}) => (
              <CourseLink
                onItemPress={onItemPressWrapper({item})}
                bottomMost={!children}
                {...item}
              />
            )}
            ref={ref => {
              if (ref)
                ref.componentDidUpdate = () => this.setState({ mounted: true });
            }}
          />
        )}
        {children && (
          <Accordion
            dataArray={children}
            expanded={!mounted && getIndexByBreadcrumb(list, expandedList, id)}
            /* first child id as list key */
            listKey={children[0].id}
            keyExtractor={item => item.id}
            style={[styles.accordion, styles.content]}
            renderContent={this._renderContent}
            renderHeader={this._renderHeader}
            ref={ref => {
              if (ref) {
                ref.setSelected = this._setSelectedWrapper(ref, id)
              }
            }}
          />
        )}
      </View>
    );
  }

  render() {
    const { list, expandedList } = this.props;
    const { mounted } = this.state;

    return (
      <Accordion
        dataArray={list}
        expanded={!mounted && getIndexByBreadcrumb(list, expandedList, '')}
        keyExtractor={item => item.id}
        style={styles.accordion}
        renderContent={this._renderContent}
        renderHeader={this._renderHeader}
        ref={ref => {
          this._rootAccordion = ref;
          if (ref)
            ref.setSelected = this._setSelectedWrapper(ref, '');
        }}
      />
    );
  }
}

CourseScrollable.defaultProps = {
  onItemPressWrapper: () => { return () => {}; },
  onSetSelectCourseIndex: () => {},
  expandedList: []
};

CourseScrollable.propTypes = {
  list: PropTypes.array.isRequired,
  onItemPressWrapper: PropTypes.func,
  onSetSelectCourseIndex: PropTypes.func,
  expandedList: PropTypes.array
};

export default CourseScrollable;
