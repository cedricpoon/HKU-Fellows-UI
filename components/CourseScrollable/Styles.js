import { StyleSheet } from 'react-native';

import { course, fontSize, gridBase } from 'hkufui/theme/grid'
import { logo } from 'hkufui/theme/palette'

export default StyleSheet.create({
  accordion: {
    borderWidth: 0
  },
  header: {
    backgroundColor: 'transparent',
    flexDirection: "row",
    padding: course.headerPadding,
    justifyContent: "space-between",
    alignItems: "center"
  },
  title: {
    flex: 0.95
  },
  content: {
    paddingLeft: course.accordionPaddingLeft,
    marginLeft: course.accordionMarginLeft,
    borderLeftWidth: 1,
    borderColor: 'lightgray'
  },
  headerText: {
    fontSize: fontSize.titleSize
  },
  exoActive: {
    color: logo.green
  },
  active: {
    color: logo.blue
  },
  linkItemNonBottomMost: {
    paddingLeft: course.headerPadding,
    marginLeft: 0,
    borderBottomWidth: 0
  },
  linkItem: {
    paddingVertical: course.headerPadding,
    paddingRight: course.headerPadding
  },
  linkDescription: {
    color: logo.black,
    fontSize: fontSize.titleSize
  },
  linkLeft: {
    flex: 0,
    paddingRight: gridBase
  },
  linkRight: {
    flex: 0.05
  },
  linkChevron: {
    color: logo.black
  }
});
