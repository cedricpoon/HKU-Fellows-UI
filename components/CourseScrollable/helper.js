export const getCoursePathByIndex = (courses, parentId, index) => {

  const transverse = (list, id) => {
    var retn = [];
    list.forEach((item) => {
      if (item.id === id) {
        if (index !== null && item.children && item.children.length > index)
          retn.unshift({
            id: item.children[index].id,
            title: item.children[index].title
          });
        retn.unshift({ id: item.id, title: item.title });
      } else if (item.children && retn.length == 0) {
        retn = transverse(item.children, id);
        if (retn.length != 0) {
          retn.unshift({ id: item.id, title: item.title });
        }
      }
    });
    return retn;
  };

  if (parentId === '') {
    return transverse([{ id: '', title: '', children:courses }], parentId).slice(1);
  } else
    return transverse(courses, parentId);
};

export const getCoursePathById = (courses, id) => {
  return getCoursePathByIndex(courses, id, null);
}
