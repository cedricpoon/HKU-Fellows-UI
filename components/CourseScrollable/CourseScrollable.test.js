import React from 'react';
import { shallow } from 'enzyme';

import CourseScrollable from './CourseScrollable';

describe('Testing CourseScrollable Component', () => {
  const defaultProps = {
    list: []
  };

  const emptyOnItemPressWrapper = () => {
    return () => {};
  };

  const emptyOnSetSelectCourseIndex = () => {};

  it('renders as expected with course list', () => {
    const wrapper = shallow(
      <CourseScrollable {...defaultProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as expected with course list and events', () => {
    const wrapper = shallow(
      <CourseScrollable
        onItemPressWrapper={emptyOnItemPressWrapper}
        onSetSelectCourseIndex={emptyOnSetSelectCourseIndex}
        {...defaultProps}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as expected with course list and default expanded settings', () => {
    const wrapper = shallow(
      <CourseScrollable {...defaultProps} expandedList={[]} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
