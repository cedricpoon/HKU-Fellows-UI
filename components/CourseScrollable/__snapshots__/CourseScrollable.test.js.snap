// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CourseScrollable Component renders as expected with course list 1`] = `
<Accordion
  dataArray={Array []}
  keyExtractor={[Function]}
  renderContent={[Function]}
  renderHeader={[Function]}
  style={
    Object {
      "borderWidth": 0,
    }
  }
/>
`;

exports[`Testing CourseScrollable Component renders as expected with course list and default expanded settings 1`] = `
<Accordion
  dataArray={Array []}
  keyExtractor={[Function]}
  renderContent={[Function]}
  renderHeader={[Function]}
  style={
    Object {
      "borderWidth": 0,
    }
  }
/>
`;

exports[`Testing CourseScrollable Component renders as expected with course list and events 1`] = `
<Accordion
  dataArray={Array []}
  keyExtractor={[Function]}
  renderContent={[Function]}
  renderHeader={[Function]}
  style={
    Object {
      "borderWidth": 0,
    }
  }
/>
`;
